### DISCUSSION TITLE
Add multiple images to channel

### DISCUSSION THREAD

**Discussion by vijaypshah** [TorchIO-project/torchio#449] [Q&A] [answered]
Hi,

I have images stored in multiple files. I would like to load them but append each image to color channel. 
I can make different type in Subject and load them. However, not sure how I can append them to color channel. 

e.g. Subject_a = tio.Subject( A=tio.ScalerImage("A.jpg"),B=tio.ScalerIamge("B.img"))
But now I would like A and B images to be color channel. Assuming that A and B images are of same size. 

Best Regards,
Vijay

**Comment by fepegar** [comment, marked as answer]
The docs for `Image` say:

> If a sequence of paths is given, data will be concatenated on the channel dimension so spatial dimensions must match.

Therefore, it should work if you do something like:

```python
subject_a = tio.Subject(image=tio.ScalarImage(["A.jpg", "B.img"]))
```

**Comment by vijaypshah** [comment]
Great Thanks! 

If my images are not in same geomerty, andwould like to first apply some transformation before concate them to channel.
Would it be something that need to be written as Lambda transformation function ?

Best Regards,
Vijay

**Reply by fepegar** [reply]
Can you tell me more about your use case? Why do you want to concatenate your images along the channel dimension? What kind of images are you using and what's the task?

**Comment by vijaypshah** [comment]
In multi- imaging sometimes the affine matrix may not be the same. Some of the example you have all images are preprocess, so I can see one could pre-process all images and then load like you indicated to concate in the channel. 

I am just experimenting as of now loading the images without having to do pre-processing in separate steps. Your package seems ideal to accomplish those tasks. 

Task:
Load image A, B, C [different dimension and affine matrix]
Make all images of same spacing and affine matrix
concate them in channel if required.

**Reply by fepegar** [reply]
Suppose you have two images from different modalities:

T1 MRI:
![Screenshot_1](https://user-images.githubusercontent.com/12688084/107160675-9f5f6100-698f-11eb-879a-af5637afc5a8.png)

T2 MRI:
![Screenshot](https://user-images.githubusercontent.com/12688084/107160680-a2f2e800-698f-11eb-8364-fa482c7de8d5.png)

These images clearly have different spacings and sizes but are registered.

```python
import torchio as tio
subject = tio.Subject(
    t1=tio.ScalarImage('t1.nii.gz'),
    t2=tio.ScalarImage('t2.nii.gz'),
)
subject.plot()
```

![Figure_1](https://user-images.githubusercontent.com/12688084/107160830-8c995c00-6990-11eb-82e6-d8b66192c2f5.png)

You can resample the second one in the space of the first:

```python
resample = tio.Resample('t1')
resampled = resample(subject)
resampled.plot()
```

![Figure_2](https://user-images.githubusercontent.com/12688084/107160836-902ce300-6990-11eb-896f-1858dcbd0198.png)

```python
In [14]: subject.t1
Out[14]: ScalarImage(shape: (1, 193, 229, 193); spacing: (1.00, 1.00, 1.00); orientation: RAS+; memory: 32.5 MiB; dtype: torch.FloatTensor)

In [15]: subject.t2
Out[15]: ScalarImage(shape: (1, 193, 37, 193); spacing: (1.00, 3.00, 1.00); orientation: RAS+; memory: 5.3 MiB; dtype: torch.FloatTensor)

In [16]: resampled.t1
Out[16]: ScalarImage(shape: (1, 193, 229, 193); spacing: (1.00, 1.00, 1.00); orientation: RAS+; memory: 32.5 MiB; dtype: torch.FloatTensor)

In [17]: resampled.t2
Out[17]: ScalarImage(shape: (1, 193, 229, 193); spacing: (1.00, 1.00, 1.00); orientation: RAS+; memory: 32.5 MiB; dtype: torch.FloatTensor)
```

You can always create your own transform to concatenate them along the channel dimension, but I would just do that before feeding them to the network:

```python
In [18]: dataset = tio.SubjectsDataset([subject], transform=resample)

In [19]: import torch

In [20]: loader = torch.utils.data.DataLoader(dataset)

In [21]: batch = iter(loader).next()

In [23]: inputs = torch.cat((batch['t1'][tio.DATA], batch['t2'][tio.DATA]), dim=1)

In [24]: inputs.shape
Out[24]: torch.Size([1, 2, 193, 229, 193])
```

**Comment by vijaypshah** [comment]
Great thank you @fepegar for detailed explaination! 
Very helpful!

Best Regards,
Vijay
